app/src/features/charts/UsaChloroplethMap.tsx

```diff
@@ -0,0 +1,228 @@
+import React, { useState, useEffect, useRef } from "react";
+import { Vega } from "react-vega";
+
+type AggregationOperation = "sum" | "mean";
+type NumberFormat = "raw" | "percentage";
+
+const HEIGHT_WIDTH_RATIO = 0.5;
+const LEGEND_WIDTH = 100;
+
+const GEO_DATASET = "GEO_DATASET";
+const GEO_ID = "id";
+
+const VAR_DATASET = "VAR_DATASET";
+const VAR_STATE_FIPS = "FIPS";
+const VAR_COUNTY_FIPS = "COUNTY_FIPS";
+
+function UsaChloroplethMap(props: {
+  dataUrl: string; // Takes CSV or JSON
+  varField: string;
+  legendTitle: string;
+  filterVar?: string;
+  filterValue?: string;
+  signalListeners: any;
+  operation?: AggregationOperation;
+  stateFips?: number;
+  numberFormat?: NumberFormat;
+}) {
+  const [width, setWidth] = useState<number | undefined>();
+  // Initial spec state is set in useEffect when default geo is set
+  const [spec, setSpec] = useState({});
+  const myRef = useRef(document.createElement("div"));
+
+  useEffect(() => {
+    /* SET UP VARIABLE DATSET */
+    let varTransformer: any[] = [];
+    // If user wants data filtered, we first apply the filter
+    if (props.filterVar && props.filterValue && props.filterValue !== "All") {
+      varTransformer.push({
+        type: "filter",
+        expr: "datum." + props.filterVar + " === '" + props.filterValue + "'",
+      });
+    }
+    // Next we perform the aggregation based on requested operation
+    if (props.operation) {
+      varTransformer.push({
+        type: "aggregate",
+        groupby: [VAR_STATE_FIPS],
+        fields: [props.varField], // field name to aggregate on
+        ops: [props.operation],
+        as: [props.varField], // field name of the aggregation
+      });
+    }
```

Comment:
I think long-term, it would be preferable to do the data manipulation outside of vega. It's easier to abstract, generalize, test, etc. It should be easier to do this once we have data coming in through the dataset loading flow, rather than reading from files, so no need to change anything right now.

Are there performance concerns (eg flickering) with doing the data manipulation outside of vega?

Reply:
Well there's already flickering now haha :)
We could probably get rid of the flickering if we wanted though.

This can help us get accurate data for UXR testing or whatever now and we can refactor later :)